ui/components-react/src/components-react/propertygrid/internal/flat-properties/FlatPropertyRenderer.tsx

```diff
@@ -104,17 +116,29 @@ export const FlatPropertyRenderer: React.FC<FlatPropertyRendererProps> = (
       );
 
     case PropertyValueFormat.Array:
-      // If array is empty, render it as a primitive property
-      if (props.propertyRecord.value.items.length === 0)
+      // If array is empty or has a custom renderer registered, render it as a primitive property
+      if (props.propertyRecord.value.items.length === 0 || hasCustomRenderer)
         return (
           <PrimitivePropertyRenderer
             highlight={highlight?.applyOnLabel ? highlight : undefined}
+            renderValueElementForNonPrimitiveProperty={hasCustomRenderer}
             {...primitiveRendererProps}
           />
         );
 
     // eslint-disable-next-line no-fallthrough
     case PropertyValueFormat.Struct:
+      // If a custom renderer is registered for the struct property, render it as a primitive property
+      if (hasCustomRenderer) {
+        return (
+          <PrimitivePropertyRenderer
+            highlight={highlight?.applyOnLabel ? highlight : undefined}
+            renderValueElementForNonPrimitiveProperty={true}
+            {...primitiveRendererProps}
+          />
```

Comment:
Maybe, instead of using `PrimitivePropertyRenderer` with special flag to render non-primitive properties, we could either - modify `FlatNonPrimitivePropertyRenderer` to take custom renderer into account,
- or, create a new "custom property renderer" that accepts only properties with custom renderer and uses it to render the property.

Reply:
When the struct property is null its type is set to primitive and so it is rendered as a primitive property, so only modifying `FlatNonPrimitivePropertyRenderer` would result in inconsistency depending if the property has a value or not.

Creating a new renderer that would handle all properties with custom renderers would work, but would end up being a duplicate of both `PrimitivePropertyRenderer` and `PropertyView` with the only difference of having the same effect as the current `renderValueElementForNonPrimitiveProperty` flag.